acceptance/restorer_test.go

```diff
@@ -0,0 +1,130 @@
+// +build acceptance
+
+package acceptance
+
+import (
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
+
+	h "github.com/buildpacks/lifecycle/testhelpers"
+)
+
+var (
+	restoreDockerContext = filepath.Join("testdata", "restorer")
+	restorerBinaryDir    = filepath.Join("testdata", "restorer", "container", "cnb", "lifecycle")
+	restorerImage        = "lifecycle/acceptance/restorer"
+)
+
+func TestRestorer(t *testing.T) {
+	h.SkipIf(t, runtime.GOOS == "windows", "Restorer acceptance tests are not yet supported on Windows")
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	h.MakeAndCopyLifecycle(t, "linux", restorerBinaryDir)
+	h.DockerBuild(t, restorerImage, restoreDockerContext)
+	defer h.DockerImageRemove(t, restorerImage)
+
+	spec.Run(t, "acceptance-restorer", testRestorer, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testRestorer(t *testing.T, when spec.G, it spec.S) {
+	when("called with arguments", func() {
+		it("errors", func() {
+			command := exec.Command("docker", "run", "--rm", restorerImage, "some-arg")
+			output, err := command.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to parse arguments: received unexpected Args"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("called without any cache arguments", func() {
+		it("outputs it will not restore cache layer data", func() {
+			command := exec.Command("docker", "run", "--rm", "--env", "CNB_CACHE_DIR=", restorerImage)
+			output, err := command.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "Not restoring cached layer data, no cache flag specified"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("using cache-dir", func() {
+		when("there is cache present from a previous build", func() {
+			var copyDir, containerName string
+
+			it.Before(func() {
+				containerName = "test-container-" + h.RandString(10)
+				var err error
+				copyDir, err = ioutil.TempDir("", "test-docker-copy-")
+				h.AssertNil(t, err)
+			})
+
+			it.After(func() {
+				if h.DockerContainerExists(t, containerName) {
+					h.Run(t, exec.Command("docker", "rm", containerName))
+				}
+				os.RemoveAll(copyDir)
+			})
+
+			it("restores cached layer data", func() {
+				h.DockerRunAndCopy(t,
+					containerName,
+					copyDir,
+					restorerImage,
+					"/layers",
+					h.WithFlags("--user", "0"),
```

Comment:
Why is this necessary?

Reply:
The `restorer` does a `chown` to the `uid`/`gid` on the `layers` and `cache` dirs before dropping privs.

Reply:
I see. But since we do that in the Dockerfile too, we could run this as `--user 1234` if we wanted right? I guess this line is sort of implying that restorer **must** be run as root, when in fact we only NEED to do that if using the docker daemon.

Reply:
Using `--user 1234` does _not_ work. If we expect that to work while using docker daemon, it does not seem to. It still attempts to `chown` the `/layers` directory even though it should already be owned by `1234`.

```
[docker run --user 1234 --name test-container-hzcuhdnfou lifecycle/acceptance/restorer -cache-dir /cache], exit status 1, , ERROR: failed to chown volumes: chown /layers: operation not permitted
```

Is this expected because we are using docker volumes?

Reply:
This surprises me. Maybe we also need to pass the group id?

Reply:
The `Dockerfile` ends with `USER 1234:1000`, so omitting the `--user` should have worked but doesn't. I did try explicitly codifying it into the test to make sure and it fails as well.
```
--- FAIL: TestRestorer/acceptance-restorer/using_cache-dir/there_is_cache_present_from_a_previous_build/restores_cached_layer_data (1.65s)
            docker.go:56: Expected nil: failed to execute command: [docker run --user 1234:1000 --name test-container-qjlelhxuou lifecycle/acceptance/restorer -cache-dir /cache], exit status 1, , ERROR: failed to chown volumes: chown /layers: operation not permitted
```

Reply:
This is interesting. So I think the disconnect here is that the restorer (and probably other phases too) when not provided a `-uid` or `-gid` flag sets those values to 0 / root instead of the user & group that the lifecycle itself is running as.

Thus when we're passing `--user 1234` (run restorer as `1234`) the lifecycle is still using `uid 0` when checking `canWrite`: https://github.com/buildpacks/lifecycle/blob/77735c391b3f7b8376616a04cad8013c46b16e59/priv/user_linux.go#L45 and thus tries to chown `/layers` even though it's already owned by `1234`. Since you can't chown as non-root it all fails...

I tried passing `-uid 1234` as an argument to the restorer and it doesn't have a problem with `/layers` but then tries to chown `/cache` :(

I think this is all tangential to what we're trying to accomplish here - maybe we could throw a comment on the `--user 0` and get around to tackling this more carefully when filling out the rest of the restorer tests. The analyzer tests currently have some unpleasant bash workarounds to get the cache directory to be readable when mounted in as a volume: https://github.com/buildpacks/lifecycle/blob/ee2fd58e6e5c16de1b450f4ddcf819849c6a6d5c/acceptance/analyzer_test.go#L464-L466

Reply:
Because `CNB_USER_ID` and `CNB_GROUP_ID` are required to be set in the build image by the spec, we can assume we always have a good value for `uid` and `gid` whether they come from flags or the environments.

Reply:
I removed all the `--user "0"` instances. Since I'm doing the tar work without volumes, I can safely do the chown in the `Dockerfile` and know that `restorer` won't have to. 😃